book/src/design/circuit/gadgets/sinsemilla.md

```diff
@@ -0,0 +1,149 @@
+
+### Layout
+
+Note: $q_{S3}$ is synthesized from $q_{S1}$ and $q_{S2}$; it is shown here only for clarity.
+
+$$
+\begin{array}{|c|c|c|c|c|c|c|c|c|c|c|c|c|}
+\hline
+\text{Step} &    x_A     &    x_P      &   bits   &    \lambda_1     &   \lambda_2      & q_{S1} & q_{S2} & q_{S3} & fixed\_y_Q & table_{idx} & table_x  & table_y  \\\hline
+    0       & x_Q        & x_{P[m_1]}  & z_0      & \lambda_{1,0}    & \lambda_{2,0}    & 1      & 1      & 0      &    y_Q     & 0           & x_{P[0]} & y_{P[0]} \\\hline
+    1       & x_{A,1}    & x_{P[m_2]}  & z_1      & \lambda_{1,1}    & \lambda_{2,1}    & 1      & 1      & 0      &     0      & 1           & x_{P[1]} & y_{P[1]} \\\hline
+    2       & x_{A,2}    & x_{P[m_3]}  & z_2      & \lambda_{1,2}    & \lambda_{2,2}    & 1      & 1      & 0      &     0      & 2           & x_{P[2]} & y_{P[2]} \\\hline
+  \vdots    & \vdots     & \vdots      & \vdots   & \vdots           & \vdots           & 1      & 1      & 0      &     0      & \vdots      & \vdots   & \vdots   \\\hline
+   n-1      & x_{A,n-1}  & x_{P[m_n]}  & z_{n-1}  & \lambda_{1,n-1}  & \lambda_{2,n-1}  & 1      & 0      & 0      &     0      & \vdots      & \vdots   & \vdots   \\\hline
+    0'      & x'_{A,0}   & x_{P[m'_1]} & z'_0     & \lambda'_{1,0}   & \lambda'_{2,0}   & 1      & 1      & 0      &     0      & \vdots      & \vdots   & \vdots   \\\hline
+    1'      & x'_{A,1}   & x_{P[m'_2]} & z'_1     & \lambda'_{1,1}   & \lambda'_{2,1}   & 1      & 1      & 0      &     0      & \vdots      & \vdots   & \vdots   \\\hline
+    2'      & x'_{A,2}   & x_{P[m'_3]} & z'_2     & \lambda'_{1,2}   & \lambda'_{2,2}   & 1      & 1      & 0      &     0      & \vdots      & \vdots   & \vdots   \\\hline
+  \vdots    & \vdots     & \vdots      & \vdots   & \vdots           & \vdots           & 1      & 1      & 0      &     0      & \vdots      & \vdots   & \vdots   \\\hline
+   n-1'     & x'_{A,n-1} & x_{P[m'_n]} & z'_{n-1} & \lambda'_{1,n-1} & \lambda'_{2,n-1} & 1      & 2      & 2      &     0      & \vdots      & \vdots   & \vdots   \\\hline
+    n'      & x'_{A,n}   &             &          & y_{A,n}          &                  & 0      & 0      & 0      &     0      & \vdots      & \vdots   & \vdots   \\\hline
+\end{array}
+$$
+
+$x_Q$, $z_0$, $z'_0$, etc. would be copied in using equality constraints.
+
+### Optimized Sinsemilla gate
+
+$\begin{array}{lrcl}
+\text{For } i \in [0, n), \text{ let} &x_{R,i} &=& \lambda_{1,i}^2 - x_{A,i} - x_{P,i} \\
+                                      &Y_{A,i} &=& (\lambda_{1,i} + \lambda_{2,i}) \cdot (x_{A,i} - x_{R,i}) \\
+                                      &y_{P,i} &=& Y_{A,i}/2 - \lambda_{1,i} \cdot (x_{A,i} - x_{P,i}) \\
+                                      &m_{i+1} &=& z_{i} - 2^k \cdot (q_{S2,i} - q_{S3,i}) \cdot z_{i+1} \\
+                                      &q_{S3}  &=& q_{S2} \cdot (q_{S2} - 1)
```

Comment:
I checked these. It's a neat trick, which I believe corresponds to a selector combining optimization (https://github.com/zcash/halo2/issues/116). Writing it out below for my own benefit.

We need two properties:
- At the end of every piece, z<sub>n-1</sub> is constrained to 10 bits, forcing z<sub>n</sub> to be zero without needing a cell.
- At the end of the last piece, we want to alter the y<sub>A,i+1</sub> constraint to constrain an output cell with y<sub>A,n</sub> instead of the next row's Sinsemilla cells.

To achieve both of these in a single selector, we define q<sub>S3</sub> in terms of q<sub>S2</sub> to get the following tri-state toggle:

| q<sub>S2</sub> |  q<sub>S3</sub> | q<sub>S2</sub> -  q<sub>S3</sub> | Piece | Position |
| --- | --- | --- | --- | --- |
| 0 | 0 | 0 | Earlier | End |
| 1 | 0 | 1 | Earlier | Within |
| 2 | 2 | 0 | Last | End |

We can now set q<sub>S2</sub> to one of three states, and use q<sub>S2</sub> -  q<sub>S3</sub> to toggle the end-of-piece behaviour, while q<sub>S3</sub> toggles the last-piece behaviour.

Reply:
Exactly right. It's basically interpolating over the values of q<sub>S2</sub>, which is also what you would do for zcash/halo2#116.